From 4e8362906ba6475bd33b37fd1290c5c39e4010bf Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Wed, 12 Apr 2023 06:34:45 +0000
Subject: [PATCH 34/53] [Huawei] mm/sharepool: Store sp_mapping in sp_area

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

We need the sp_mapping for sp_area in sp_area_free(). We cannot get the
sp_mapping from spa->spg because the sp_group may have been destroyed
before sp_area_free().

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 43 +++++++++++++++----------------------------
 1 file changed, 15 insertions(+), 28 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index ffef24559107..0bcdcb4ff1ee 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -486,18 +486,6 @@ static int sp_mapping_group_setup(struct mm_struct *mm, struct sp_group *spg)
 	return 0;
 }
 
-static struct sp_mapping *sp_mapping_find(struct sp_group *spg,
-						 unsigned long addr)
-{
-	if (addr >= MMAP_SHARE_POOL_NORMAL_START && addr < MMAP_SHARE_POOL_NORMAL_END)
-		return spg->mapping[SP_MAPPING_NORMAL];
-
-	if (addr >= MMAP_SHARE_POOL_RO_START && addr < MMAP_SHARE_POOL_RO_END)
-		return spg->mapping[SP_MAPPING_RO];
-
-	return spg->mapping[SP_MAPPING_DVPP];
-}
-
 static struct sp_group *create_spg(int spg_id);
 static void free_new_spg_id(bool new, int spg_id);
 static void free_sp_group_locked(struct sp_group *spg);
@@ -708,7 +696,8 @@ struct sp_area {
 	struct rb_node rb_node;		/* address sorted rbtree */
 	struct rb_node spg_link;	/* link to the spg->rb_root */
 	struct sp_group *spg;
-	enum spa_type type;		/* where spa born from */
+	struct sp_mapping *spm;		/* where spa born from */
+	enum spa_type type;
 	unsigned long kva;		/* shared kva */
 	pid_t applier;			/* the original applier process */
 	int preferred_node_id;		/* memory node */
@@ -1715,18 +1704,19 @@ static struct sp_area *sp_area_alloc(unsigned long size, unsigned long flags,
 		goto error;
 	}
 
-	spa->va_start = addr;
-	spa->va_end = addr + size_align;
-	spa->real_size = size;
-	spa->region_vstart = vstart;
-	spa->flags = flags;
-	spa->is_hugepage = (flags & SP_HUGEPAGE);
-	spa->spg = spg;
-	atomic_set(&spa->use_count, 1);
-	spa->type = type;
-	spa->kva = 0;   /* NULL pointer */
-	spa->applier = applier;
+	spa->va_start          = addr;
+	spa->va_end            = addr + size_align;
+	spa->real_size         = size;
+	spa->region_vstart     = vstart;
+	spa->flags             = flags;
+	spa->is_hugepage       = (flags & SP_HUGEPAGE);
+	spa->spg               = spg;
+	spa->spm               = mapping;
+	spa->type              = type;
+	spa->kva               = 0;   /* NULL pointer */
+	spa->applier           = applier;
 	spa->preferred_node_id = node_id;
+	atomic_set(&spa->use_count, 1);
 
 	/* the link location could be saved before, to be optimized */
 	spm_insert_area(mapping, spa);
@@ -1789,11 +1779,8 @@ static bool vmalloc_area_clr_flag(unsigned long kva, unsigned long flags)
  */
 static void sp_area_free(struct sp_area *spa)
 {
-	unsigned long addr = spa->va_start;
-	struct sp_mapping *spm;
+	struct sp_mapping *spm = spa->spm;
 
-	/* FIXME: UAF to be fix */
-	spm = sp_mapping_find(spa->spg, addr);
 	spin_lock(&spm->sp_mapping_lock);
 	if (spm->free_area_cache) {
 		struct sp_area *cache;
-- 
2.17.1

